### PR TITLE
Speed up turbine loading operation

### PR DESCRIPTION
**Pull request by misi9170**
@Bartdoekemeijer points out in #965 that instantiating the `FlorisModel` can be slow, especially when there are many turbines. @achenry had also pointed out a similar problem in #885. 

After looking into this some, and using the profiling method suggested by @achenry, I found that a lot of time is being spent in the `construct_turbine_map()` method of `Farm`. Essentially, each `Turbine` is being instantiated separately, even if all turbines are the same type (which is often the case). 

This PR proposes a fix to this that only calls `Turbine.from_dict()` once for each _type_ of turbine, and then uses those (fewer) instantiations to construct the `Farm.turbine_map`. I'm seeing a substantial speed up, and now I think (after the change) that instantiation speed is dominated by `n_findex` rather than `n_turbines`.

There may be other opportunities for speed up during instantiation, but I think this may be the lowest hanging fruit.

### Notes
I took the opportunity to also partly address #864, which raises the issue that the `"turbine_type"` key of the turbine dictionary _must_ be changed in order to implement changes to other parts of the dictionary. While the changes I have made here do not necessarily "fix" the issue, now an informative error is raised when this situation arises.

### Affected areas of the code
- floris/core/farm.py

### Tests
- All existing tests pass
- New test for checking that an error is raised if two different turbine definitions are given with the same `"turbine_type"`
- New test that checks that modified turbine definitions make it into the `turbine_map` correctly

### Examples
The main relevant example is examples/examples_turbine/002_multiple_turbine_types.py. I see no change to the outputs, as expected.

### Performance
Using the example code in #965, I get the following performance on the develop branch:
dis965

while this PR provides:
dis965_fix

### CODE REVIEW

**Comment by paulf81** [comment]
Here is also the wrote from running it on python 3.11 on my machine:

image

**Comment by misi9170** [comment]
Looking across the speed results here and on #965, it seems that the "remaining" time between the n_findex=300 and n_findex=3000 cases is approximately 5 seconds across various machines (or more like 3 seconds when using python 3.11) when setting 1000 turbines.

**Comment by misi9170** [comment]
> Reviewing the code and checking the tests this all looks good. the only thought I had, would 1 or 2 additional tests where the number of unique turbine types is different than the number of turbines (and some assertions about what should be observed) be good to add while we're at it? Or is that all sufficiently covered by existing tests?

Yeah, I could add a quick test for that. The test I've added currently have 5 turbines, one of which is different, but I should check that now if `"turbine_type"` _is_ renamed, the setup goes correctly.

**Comment by misi9170** [comment]
Slight update on the timings where we aren't rounding the time to the nearest integer:
dis965_fix_float

**Comment by misi9170** [comment]
> Reviewing the code and checking the tests this all looks good. the only thought I had, would 1 or 2 additional tests where the number of unique turbine types is different than the number of turbines (and some assertions about what should be observed) be good to add while we're at it? Or is that all sufficiently covered by existing tests?

I have now added this test

**Comment by paulf81** [comment]
Good to go from me then!

**Comment by Bartdoekemeijer** [comment]
Awesome!

Here's my performance with 3.11.8 on WSL with Ubuntu 22 LTS:
image

Thank you so much for fixing this, and especially on such short notice!!

**Comment by paulf81** [comment]
> Awesome!
> 
> Here's my performance with 3.11.8 on WSL with Ubuntu 22 LTS: image
> 
> Thank you so much for fixing this, and especially on such short notice!!

A nice improvement!

File: floris/core/farm.py [hunk 1]
```diff
@@ -134,21 +138,29 @@ def __attrs_post_init__(self) -> None:
         # This allows to read the yaml input files once rather than every time they're given.
         # In other words, if the turbine type is already in the cache, skip that iteration of
         # the for-loop.
+
         turbine_definition_cache = {}
```
Comment (bayc) [inline floris/core/farm.py:142]: Could this just be modified to be `self._turbine_definition_cache`? Then you could remove line 219, I believe. Also, update `turbine_definition_cache` to `self._turbine_definition_cache` throughout.

**Review by bayc — APPROVED** [review]

**Comment by bayc** [comment]
@misi9170 Yea, this looks nice! Just one suggestion in my comment above.

**Comment by misi9170** [comment]
> @misi9170 Yea, this looks nice! Just one suggestion in my comment above.

This is now done

**Pull request merged**